Consolidate RID and native file naming in MSBuild scripts

am11 opened the pull request:
* Use short variable names for native files naming convention, that are
  used by `framework.sharedfx.targets` in arcade, and cleanup
  redefinitions from crossgen2 and installer. e.g. `ExeSuffix` instead
  of `ApplicationFileExtension`, `LibSuffix` instead of
  `LibraryFileExtension` and so on.
* Calculate `TargetArchitecture`, `NonPortableRuntimeOS` (for
  `PortableBuild`) and `PackageRID` values once for the entire
  livebuild, inside `eng/Configurations.props`. This implementation is
  a union of three varied implementations that are being deleted.
* Import `names.props` once in `eng/Configurations.props` based on
  calculated `PackageRID` and cleanup imports of this file from various
  places.
* Combine OS targets definition in MSBuild scripts.

am11 commented:
cc @janvorli

am11 commented on eng/Configurations.props:
```diff
@@ -68,6 +70,92 @@
+    <_runtimeOSFamily Condition="'$(_runtimeOSVersionIndex)' != '-1'">$(_runtimeOS.SubString(0, $(_runtimeOSVersionIndex)))</_runtimeOSFamily>
+
+    <_buildingInOSX>$([MSBuild]::IsOSPlatform('OSX'))</_buildingInOSX>
+    <_portableOS>linux</_portableOS>
+    <_portableOS Condition="'$(_runtimeOS)' == 'linux-musl'">linux-musl</_portableOS>
+    <_portableOS Condition="$(_buildingInOSX)">osx</_portableOS>
+    <_portableOS Condition="'$(_runtimeOSFamily)' == 'win' or '$(TargetOS)' == 'Windows_NT'">win</_portableOS>
+    <_portableOS Condition="'$(_runtimeOSFamily)' == 'FreeBSD'">freebsd</_portableOS>
+    <_portableOS Condition="'$(_runtimeOSFamily)' == 'illumos'">illumos</_portableOS>
+    <_portableOS Condition="'$(_runtimeOSFamily)' == 'Solaris'">solaris</_portableOS>
+    <_portableOS Condition="'$(_runtimeOS)' == 'Browser'">browser</_portableOS>
+    <_portableOS Condition="'$(_runtimeOS)' == 'ios'">ios</_portableOS>
+    <_portableOS Condition="'$(_runtimeOS)' == 'tvos'">tvos</_portableOS>
+    <_portableOS Condition="'$(_runtimeOS)' == 'android'">android</_portableOS>
```
@dagood, I was thinking that maybe it would be a good idea to move this flat list and/or `<_outputRID>` below to arcade repo in the future, in order to DRY the portable platform matrix for https://github.com/dotnet/sdk/issues/14296#issuecomment-715613987 (and other similar use-cases originating from arcade-powered repos). This will make new platforms bringup slightly smoother (without needing to update many repos individually).

In this context:
* `<PackageRID>` is the identifier used to restore packages. In cross-building scenarios (such as; building FreeBSD bits on Linux), the value of OS is `linux` to make package restore work correctly. The final output location in coreclr/libraries/tests are determined by host OS (`freebsd` in case of `FreeBSD`), which is a separate property defined elsewhere.
* `<OutputRid>` is the identifier used to determine the output directory name of installer (among other things). This is a different than how it is done for coreclr/libraries/tests, and can be simplified/removed in the future; after aligning installer with other components of runtime repo.

dagood replied:
I agree, centralizing this would be great! I remember that this has come up before, but it seemed unapproachable (at the time) due to the variety of ways RIDs are handled in various repos, interactions with native builds, etc. Unfortunately, I'm not able to find records of any of that or who was involved... I think the next step would be to file a dotnet/arcade issue to get input from the various people we'd want to start depending on this as an arcade feature.

Aside from how source-build needs to be able to override it, though, I don't think I'll be able to get hugely involved... I think pinging dotnet/runtime infra area owners and folks involved in SDK infra would be a reasonable start. (Not doing so here in case you want to keep this PR focused on this more local centralization for now.)

am11 replied:
Thanks for the tips on follow up. Yes, I would like to keep this PR focused on local centralization (internal refactoring, without changing the public contract). 

There is some follow up work required to bring about simplifications. Currently, there are multiple ways to override RID from environment variables; build scripts can influence the value of `__DistroRid` based on what was passed to `-os` flag, MSBuild property `RuntimeOS` and then `OutputRid` which is a global knob only used by installer (PR is removing its dependency from coreclr). `OutputRid` is a one-off concept, which is conditionally set to `RuntimeIdentifier`, `OfficialRidBuildRID` etc. and makes the repo-wide RID story hard to reason about. source-build is also overriding `OutputRid`, in addition to supplying `-os` as well as setting `PortableBuild` to `false`. 🤯

If we can get rid of `OutputRid` soon and agree on just one way to override the RID (via `./build.sh -os ...` which is implicitly connected to `/p:RuntimeOS=...`, when MSBuild is available on the platform), that will make this inherently-complex concept relatively easier to deal with, IMO.

am11 commented:
cc @ViktorHofer, @jkoritzinsky, @janvorli,
PTAL, resolved merge conflicts and rebased. Both CI crashes in System.Runtime.Serialization.Formatters.Tests are #44250.

ViktorHofer commented:
@jkoritzinsky kindly asking for a review

jkoritzinsky commented:
I will review this as soon as my shared framework PR is merged in. I think there will likely be conflicts between the two PRs that will have to be resolved, but I'll do a pass though just to be sure.

am11 commented:
Rebased and resolved merge conflicts with shared framework changes.

am11 commented:
Resolved merge conflicts again. really hope this is the last time for this PR, as retesting it on array of machines is no fun.. 😁

ViktorHofer commented:
Thanks a lot @am11. Jeremy will take a look soon :)

am11 commented:
@jkoritzinsky, @ViktorHofer, with this change, we are left with two flat lists where `Targets<PlatformName>` properties are set (instead of five in the base branch); one in `eng/Configurations.props` (for non-libraries components) and the other in `eng/targetframeworksuffix.props` (for libraries; that dials into SDK's TFM hook). I have added `SkipInferTargetOSName` in libraries' `Directory.Build.props` to prevent libs using these properties from `Configurations`. If it is possible, it would help to de-duplicate and flatten out these two lists as well in the future.

am11 committed:
Consolidate RID and native file naming in MSBuild scripts
* Use short variable names for native files naming convention, that are
  used by `framework.sharedfx.targets` in arcade, and cleanup
  redefinitions from crossgen2 and installer. e.g. `ExeSuffix` instead
  of `ApplicationFileExtension`, `LibSuffix` instead of
  `LibraryFileExtension` and so on.
* Calculate `TargetArchitecture`, `NonPortableRuntimeOS` (for
  `PortableBuild`) and `PackageRID` values once for the entire
  livebuild, inside `eng/Configurations.props`. This implementation is
  a union of three varied implementations that are being deleted.
* Import `names.props` once in `eng/Configurations.props` based on
  calculated `PackageRID` and cleanup imports of this file from various
  places.
* Combine OS targets definition in MSBuild scripts.

jkoritzinsky commented on src/installer/Directory.Build.props:
```diff
@@ -138,175 +115,22 @@
-    <TargetsSles>false</TargetsSles>
-    <TargetsBrowser>false</TargetsBrowser>
+  <PropertyGroup Condition="'$(PortableBuild)' == 'true'">
+    <HostMachineRidTargetsDebianPackages Condition="
+      $(NonPortableRuntimeOS.StartsWith('debian')) or
+      $(NonPortableRuntimeOS.StartsWith('ubuntu')) or
+      $(NonPortableRuntimeOS.StartsWith('linuxmint'))">true</HostMachineRidTargetsDebianPackages>
+    <!-- If the build machine isn't known to be Debian-based, try to build RPM packages. -->
+    <HostMachineRidTargetsRpmPackages Condition="'$(HostMachineRidTargetsDebianPackages)' != 'true'">true</HostMachineRidTargetsRpmPackages>
```
The tooling that the installer build uses doesn't use these properties any more. If there's no explicit usage of these properties in the repo, you can delete this PropertyGroup

jkoritzinsky commented on src/installer/Directory.Build.props:
```diff
@@ -138,175 +115,22 @@
 
   <PropertyGroup>
     <CompressedFileExtension Condition="'$(TargetOS)' == 'windows'">.zip</CompressedFileExtension>
     <CompressedFileExtension Condition="'$(TargetOS)' != 'windows'">.tar.gz</CompressedFileExtension>
     <InstallerExtension Condition="'$(TargetOS)' == 'windows'">.msi</InstallerExtension>
     <InstallerExtension Condition="'$(TargetOS)' == 'OSX'">.pkg</InstallerExtension>
-    <InstallerExtension Condition="'$(TargetsDebian)' == 'true' or '$(TargetsUbuntu)' == 'true' or '$(TargetsLinuxMint)' == 'true' or '$(HostMachineRidTargetsDebianPackages)' == 'true'">.deb</InstallerExtension>
-    <InstallerExtension Condition="'$(TargetsRhel)' == 'true' or '$(TargetsCentos)' == 'true' or '$(TargetsOpensuse)' == 'true' or '$(TargetsFedora)' == 'true' or '$(TargetsOracle)' == 'true' or '$(TargetsSles)' == 'true' or '$(HostMachineRidTargetsRpmPackages)' == 'true'">.rpm</InstallerExtension>
+    <InstallerExtension Condition="$(OutputRid.StartsWith('debian')) or $(OutputRid.StartsWith('ubuntu')) or $(OutputRid.StartsWith('linuxmint')) or '$(HostMachineRidTargetsDebianPackages)' == 'true'">.deb</InstallerExtension>
+    <InstallerExtension Condition="$(OutputRid.StartsWith('rhel')) or $(OutputRid.StartsWith('centos')) or $(OutputRid.StartsWith('opensuse')) or $(OutputRid.StartsWith('fedora')) or $(OutputRid.StartsWith('oracle')) or $(OutputRid.StartsWith('sles')) or '$(HostMachineRidTargetsRpmPackages)' == 'true'">.rpm</InstallerExtension>
     <CombinedInstallerExtension Condition="'$(TargetOS)' == 'windows'">.exe</CombinedInstallerExtension>
     <CombinedInstallerExtension Condition="'$(TargetOS)' != 'windows'">$(InstallerExtension)</CombinedInstallerExtension>
```
Same as above with this block. These properties are handled within the new tooling so we don't need to set them here.

am11 replied:
These properties were used in an unused tarter `GetInstallerLocations`. I have deleted it as well.

jkoritzinsky commented on src/tests/build.cmd:
```diff
@@ -560,7 +553,6 @@ echo Build type: one of Debug, Checked, Release ^(default: Debug^).
 echo skipmanaged: skip the managed tests build
 echo skipnative: skip the native tests build
 echo skiprestorepackages: skip package restore
-echo runtimeid ^<ID^>: Builds a test overlay for the specified OS ^(Only supported when building against packages^). Supported IDs are:
```
The indented lines below this echo can also be removed.

jkoritzinsky reviewed (approved):
Just a few comments around a few more things that can be deleted.

Other than that, LGTM if CI is green.

am11 committed:
Delete legacy tooling properties

am11 committed:
Delete legacy tooling properties

am11 commented:
windows x86 failure is https://github.com/dotnet/runtime/issues/43389.

am11 commented:
Thanks for the review @jkoritzinsky. I have addressed the CR feedback. Could this be merged?

jkoritzinsky commented:
I'm rerunning the failed job just to validate that the failure was a flaky test. Once that's green, I'll merge.
